tests/servers/server1/main.go

```diff
@@ -252,3 +263,29 @@ func handleEmbeddedResource(
 		},
 	}, nil
 }
+
+// tool404Handler intercepts requests for the always_404 tool and returns 404
+type tool404Handler struct {
+	handler http.Handler
+}
+
+func (h *tool404Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "POST" {
+		body, err := io.ReadAll(r.Body)
+		if err == nil && bytes.Contains(body, []byte("always_404")) {
+			log.Printf("Intercepting always_404 tool call - returning HTTP 404")
+			r.Body = io.NopCloser(bytes.NewReader(body))
+
+			sessionID := r.Header.Get("mcp-session-id")
+			if sessionID != "" {
+				w.Header().Set("mcp-session-id", sessionID)
+			}
```

Comment:
Is setting the session id in a response header for debugging purposes?
Looking at the protocol spec, https://modelcontextprotocol.io/specification/2025-06-18/basic/transports#session-management, the 404 status code is the only 'must' `after which it MUST respond to requests containing that session ID with HTTP 404 Not Found`.

I think it's fine if it doesn't cause any problems though